universum/modules/structure_handler.py

```diff
@@ -197,15 +198,22 @@ def process_one_step(self, merged_item: Step, step_executor: Callable, skip_exec
             self.report_skipped_block(numbering + "'" + merged_item.name + "'")
             return True
 
+        process = None
+        executed_successfully: bool = True
         # Here pass_errors=False, because any exception while executing build step
         # can be step-related and may not affect other steps
         with self.block(block_name=step_label, pass_errors=False):
-            error: Optional[str] = self.execute_one_step(merged_item, step_executor)
-            if error is not None:
+            process= self.execute_one_step(merged_item, step_executor)
+            error = process.get_error()
+            executed_successfully = (error is None)
+            if not executed_successfully:
+                error = error if error else ""
```

Comment:
In what case "error" could be empty string?

Reply:
This was a way to fix
```
universum/modules/structure_handler.py:213: error: Argument 1 to "fail_current_block" of "StructureHandler" has incompatible type "Optional[str]"; expected "str"
```
error on 
```
self.fail_current_block(error)
```
line. 
I've added mypy "ignore" tag instead.

Reply:
No, my question was not about None value, it was about empty string.
There are 3 distinct cases for values of the error variable in your code:
1. error is None
2. error is ""
3. error is some non-empty string

The line `if not executed_successfully` checks for cases 1 vs 2,3.
The line `error = error if error else ""` checks for cases 1,2 vs 3.
Considering that `error = error if error else ""` is inside the if block, the only case when it matters is case 2, the `error` being empty string.

Overall this block of changes makes an impression that empty error string is somehow important, while AFAIK it is not even possible to get one from the launcher module.

That's why I asked the original question.

Reply:
Ok, the current version is much easier to understand.
However, now I'm wondering why mypy cannot figure out that error is not None inside `if not executed_successfully` block.

Reply:
Probably that is because the type of error variable is declared explicitly.